File: A2/Sem4/MPP/Teme/BileteJavaSocketsJson/BileteServer/src/main/java/bilete/server/BileteServerImpl.java
package bilete.server;

import bilete.domain.Show;
import bilete.domain.Ticket;
import bilete.domain.User;
import bilete.services.BileteException;
import bilete.services.IBileteObserver;
import bilete.services.IBileteServices;
import repository.ShowRepository;
import repository.TicketRepository;
import repository.UserRepository;

import java.time.LocalDate;
import java.util.Collection;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class BileteServerImpl implements IBileteServices {
    private UserRepository userRepository;
    private ShowRepository showRepository;
    private TicketRepository ticketRepository;
    private Map<String, IBileteObserver> loggedClients;

    public BileteServerImpl(UserRepository userRepository, ShowRepository showRepository, TicketRepository ticketRepository) {
        this.userRepository = userRepository;
        this.showRepository = showRepository;
        this.ticketRepository = ticketRepository;
        this.loggedClients = new ConcurrentHashMap<>();
    }

    public synchronized void login(User user, IBileteObserver client) throws BileteException {
        User userR=userRepository.login(user.getUsername(), user.getPassword());
        if (userR!=null){
            if(loggedClients.get(user.getUsername())!=null)
                throw new BileteException("User already logged in.");
            loggedClients.put(user.getUsername(), client);
        }else
            throw new BileteException("Authentication failed.");
    }

    public synchronized void logout(User user, IBileteObserver client) throws BileteException {
        IBileteObserver localClient=loggedClients.remove(user.getUsername());
        if (localClient==null)
            throw new BileteException("User "+user.getId()+" is not logged in.");
    }

    public Show[] findAllArtisti() {
        Collection<Show> shows = showRepository.findAll();
        return shows.toArray(new Show[shows.size()]);
    }

    public Show[] findShowsOnDate(LocalDate date) {
        Collection<Show> shows = showRepository.findByDay(date);
        return shows.toArray(new Show[shows.size()]);
    }


    public synchronized void sellTicket(Ticket ticket) throws BileteException {
        if(ticket.getShow() == null || ticket.getShow().getArtist() == null ||
        ticket.getCostumerName() == null || ticket.getCostumerName().length() == 0 || ticket.getSeats() <= 0)
            throw new BileteException("Ticket invalid!");
        if(ticket.getShow().getAvailableSeats() < ticket.getSeats())
            throw new BileteException("Not enough seats!");
        ticketRepository.save(ticket);
        Show show = ticket.getShow();
        show.setAvailableSeats(show.getAvailableSeats() - ticket.getSeats());
        show.setSoldSeats(show.getSoldSeats() + ticket.getSeats());
        showRepository.update(show.getId(), show);

        System.out.println("Notifying clients...");
        //notifyClientsExecutor(ticket);
        for(IBileteObserver client : loggedClients.values()) {
            if (client != null) {
                try {
                    System.out.println("Notifying client " + client);
                    client.bileteUpdated(ticket);
                }
                catch (BileteException e) {
                    System.err.println("Error notifying client " + e);
                }// sending response
            }
        }
    }

    private final int defaultThreadsNo=5;
    private synchronized void notifyClientsExecutor(Ticket ticket) {
        ExecutorService executor= Executors.newFixedThreadPool(defaultThreadsNo);
        for(IBileteObserver client : loggedClients.values()) {
            if (client != null)
                executor.execute(() -> {
                    try {
                        System.out.println("Notifying client " + client);
                        System.out.println("Client class" + client.getClass().getName());
                        client.bileteUpdated(ticket);
                    } catch (BileteException e) {
                        System.err.println("Error notifying client " + e);
                    }
                });
        }
    }
}

File: A2/Sem4/MPP/Teme/Auxiliare/TemaCsNetw2/Server/BileteServerImpl.cs
using bilete.model;
using bilete.persistence;
using bilete.services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace chat.server
{
    public class BileteServerImpl : IBileteServices
    {
        private IUserRepository userRepository;
        private IShowRepository showRepository;
        private ITicketRepository ticketRepository;
        private readonly IDictionary<string, IBileteObserver> loggedClients;

        public BileteServerImpl(IUserRepository userRepository, IShowRepository showRepository, ITicketRepository ticketRepository)
        {
            this.userRepository = userRepository;
            this.showRepository = showRepository;
            this.ticketRepository = ticketRepository;
            this.loggedClients = new Dictionary<string, IBileteObserver>();
        }

        public void login(User user, IBileteObserver client)
        {
            User userR = userRepository.Find(user.Username, user.Password);
            if (userR != null)
            {
                if (loggedClients.ContainsKey(user.Username))
                    throw new BileteException("User already logged in!");
                loggedClients[user.Username] = client;
            }
            else 
                throw new BileteException("Authentication failed.");
        }

        public void logout(User user, IBileteObserver client)
        {
            IBileteObserver localClient = loggedClients[user.Username];
            if (localClient == null)
                throw new BileteException($"User {user.Username} is not logged in!");
            loggedClients.Remove(user.Username);
        }

        public Show[] findAllArtisti()
        {
            IEnumerable<Show> shows = showRepository.FindAll();
            return shows.ToArray();
        }

        public Show[] findShowsOnDate(DateOnly date)
        {
            IEnumerable<Show> shows = showRepository.FindByDay(date);
            return shows.ToArray();
        }

        public void sellTicket(Ticket ticket)
        {
            if (ticket.Show == null || ticket.Show.Artist == null || ticket.CostumerName == null || ticket.CostumerName.Length == 0 || ticket.Seats <= 0)
            {
                throw new BileteException("Ticket invalid!");
            }
            if (ticket.Show.AvailableSeats < ticket.Seats)
            {
                throw new BileteException("Not enough seats!");
            }
            ticketRepository.Add(ticket);
            ticket.Show.AvailableSeats -= ticket.Seats;
            ticket.Show.SoldSeats += ticket.Seats;
            showRepository.Update(ticket.Show.ID, ticket.Show);

            Console.WriteLine("Notifying clients ...");
            foreach(IBileteObserver client in loggedClients.Values)
                if(client != null)
                {
                    try
                    {
                        Console.WriteLine("Notifying client: " + client);
                        Task.Run(() => {
                            client.bileteUpdated(ticket);
                            Console.WriteLine($"Client {client} notified.");
                        });
                    }
                    catch(BileteException ex)
                    {
                        Console.WriteLine("Error notifying client " + client + ": " + ex.StackTrace);
                    }
                }
        }
    }
}
